sdk/servicebus/azure-servicebus/azure/servicebus/_servicebus_receiver.py

```diff
@@ -182,14 +182,45 @@ def _open(self):
 
     def _receive(self, max_batch_size=None, timeout=None):
         # type: (Optional[int], Optional[float]) -> List[ReceivedMessage]
+        # pylint: disable=protected-access
         self._open()
-        max_batch_size = max_batch_size or self._handler._prefetch  # pylint: disable=protected-access
 
+        amqp_receive_client = self._handler
+        received_messages_queue = amqp_receive_client._received_messages
+        max_batch_size = max_batch_size or self._prefetch
         timeout_ms = 1000 * (timeout or self._idle_timeout) if (timeout or self._idle_timeout) else 0
-        batch = self._handler.receive_message_batch(
-            max_batch_size=max_batch_size,
-            timeout=timeout_ms
-        )
+        abs_timeout_ms = amqp_receive_client._counter.get_current_ms() + timeout_ms if timeout_ms else 0
+
+        batch = []
+        while not received_messages_queue.empty() and len(batch) < max_batch_size:
+            batch.append(received_messages_queue.get())
+            received_messages_queue.task_done()
+        if len(batch) >= max_batch_size:
+            return batch
+
+        # Dynamically issue link credit if max_batch_size > 1 when the prefetch is the default value 1
+        if max_batch_size and self._prefetch == 1 and max_batch_size > 1:
+            link_credit_needed = max_batch_size - len(batch)
+            amqp_receive_client.message_handler.reset_link_credit(link_credit_needed)
```

Comment:
Do we need to check for a max/min link credit value before calling this?

Reply:
(I'm curious what happens if this gets passed "-1" or "100000"

Reply:
I don't think we need.

for the min (negative value) case: the prior comparison between `len(batch)` and `max_batch_size` guarantees that `link_credit_needed` is > 0. (and the underlying C code only accepts `uint` type for link credit`

for the max (very big value) case: I don't see limitation from the service doc. I've tested EventHub, the link credit can be as large as the `uint` max -- 4294967295.